pandas/core/index.py

```diff
@@ -3299,6 +3316,425 @@ def _wrap_joined_index(self, joined, other):
+            new_start = cls._ensure_int(new_start)
+            new_stop = cls._ensure_int(new_stop)
+            new_step = cls._ensure_int(new_step)
+            if new_step == 0:
+                raise ValueError("Step must not be zero")
+            #assert len(kwargs) == 0
+            return cls._simple_new(new_start, new_stop, new_step, name)
+        except TypeError:
+            # pass all invalid inputs to Int64Index to handle
+            if step is None:
+                step = False
+            return super(RangeIndex, cls).__new__(cls.__bases__[0],
+                                                  data=start,
+                                                  dtype=stop,
+                                                  copy=step,
+                                                  name=name,
+                                                  fastpath=fastpath,
+                                                  **kwargs)
+
+    @classmethod
+    def _simple_new(cls, start, stop, step, name=None):
+        result = object.__new__(cls)
+        result._start = start
+        result._stop = stop
+        result._step = step
+        result.name = name
+        result.is_unique = True
+        return result
+
+    @classmethod
+    def _ensure_int(cls, value):
+        try:
+            int_value = int(value)
+            # don't allow casting 1-element arrays to int!
+            if int_value != value or hasattr(value, '__len__'):
+                raise Exception
+        except Exception:
+            raise TypeError("Need to pass integral values")
+        return int_value
+
+    @cache_readonly
+    def _data(self):
+        return np.arange(self.start, self.stop, self.step, dtype=np.int64)
+
+    @cache_readonly
+    def _int64index(self):
+        return Int64Index(self._data, name=self.name, fastpath=True)
+
+    @property
+    def dtype(self):
+        return np.dtype(np.int64)
+
+    @property
+    def start(self):
+        return self._start
+
+    @property
+    def stop(self):
+        return self._stop
+
+    @property
+    def step(self):
+        return self._step
+
+    #@cache_readonly(allow_setting=True)
+    #def is_unique(self):
+    #    """ return if the index has unique values """
+    #    return True
+
+    #@property
+    #def has_duplicates(self):
+    #    return not self.is_unique
+
+    def tolist(self):
+        return list(range(self.start, self.stop, self.step))
+
+    def _shallow_copy(self, values=None, **kwargs):
+        """ create a new Index, don't copy the data, use the same object attributes
+            with passed in attributes taking precedence """
+        if values is None:
+            return RangeIndex(self.start, self.stop, self.step, 
+                              name=self.name, fastpath=True)
+        else:
+            name = kwargs.get('name', self.name)
+            return self._int64index._shallow_copy(values, **kwargs)
```

Comment:
Of course, thank you! Now I am finally getting somewhere with this. Only remaining issues are with `view` and `__getitem__()` using slices - which you already commented on on the other PR.